File: algorithm/src/services/soundcloudService.js
/**
 * Retrieves candidate tracks for a genre, falling back to mock data if the API isn't configured or unavailable
 * Uses SoundCloud API v2, so response fields may change over time. 
 */

const { byGenre: mockSongsByGenre } = require('../data/mockSongs');
 
const DEFAULT_LIMIT = 20;
 
/**
 * Maps raw SoundCloud API response to the application's track format
 * @param {object} raw - raw track from SoundCloud API
 * @returns {object} normalized track object
 */
function trackFromApi(raw) {
  return {
    id: raw.id,
    title: raw.title,
    artist: raw.user && raw.user.username,
    genre: raw.genre,
    duration_ms: raw.duration,
    permalink_url: raw.permalink_url,
    artwork_url: raw.artwork_url,
    streamable: raw.streamable,
    policy: raw.policy,
  };
}


/**
 * Fetches tracks matching genre. Returns mock data to continue operations if Soundcloud client ID is unavailable
 * @param {string} genre - The genre to search for
 * @param {object} [options] 
 * @param {number} [options.limit=20] - Max number of tracks to return
 * @returns {Promise<Array<object>>} List of tracks matching the genre
 */
async function getTracksByGenre(genre, { limit = DEFAULT_LIMIT } = {}) {
  const clientId = process.env.SOUNDCLOUD_CLIENT_ID;
 
  if (!clientId) {
    console.warn('[soundcloudService] SOUNDCLOUD_CLIENT_ID not set — using mock songs');
    return mockSongsByGenre(genre);
  }
 
  const url = new URL('https://api-v2.soundcloud.com/search/tracks');
  url.searchParams.set('q', genre);
  url.searchParams.set('client_id', clientId);
  url.searchParams.set('limit', String(limit));
 
  try {
    const res = await fetch(url);
    if (!res.ok) {
      throw new Error(`SoundCloud responded with status ${res.status}`);
    }
    const data = await res.json();
    const tracks = (data.collection || []).map(trackFromApi);
    return tracks.length ? tracks : mockSongsByGenre(genre);
  } catch (err) {
    console.warn(`[soundcloudService] SoundCloud request failed (${err.message}) — falling back to mock songs`);
    return mockSongsByGenre(genre);
  }
}
 
module.exports = { getTracksByGenre };

File: algorithm/src/services/soundcloudService.py
"""
 * Retrieves candidate tracks for a genre, falling back to mock data if the API isn't configured or unavailable
 * Uses SoundCloud API v2, so response fields may change over time. 
"""
import os
import re
import requests
from soundcloud import SoundCloud
 
from src.data.mockSongs import by_genre as mock_songs_by_genre
 
DEFAULT_LIMIT = 20
 
_client_id = None
 
def _get_client_id():
    global _client_id
    if _client_id is None:
        _client_id = os.environ.get('SOUNDCLOUD_CLIENT_ID') or SoundCloud.generate_client_id()
    return _client_id
 
 
def _redact(text):
    """SoundCloud error strings embed the full request URL, client_id
    included -- strip it out before this ever hits a log."""
    return re.sub(r'client_id=[^&\s]+', 'client_id=***', str(text))
 
 
def _track_from_api(raw):
    return {
        'id': raw.get('id'),
        'title': raw.get('title'),
        'artist': (raw.get('user') or {}).get('username'),
        'genre': raw.get('genre'),
        'duration_ms': raw.get('duration'),
        'permalink_url': raw.get('permalink_url'),
        'artwork_url': raw.get('artwork_url'),
        'streamable': raw.get('streamable'),
        'policy': raw.get('policy'),
    }
 
def get_tracks_by_genre(genre, limit=DEFAULT_LIMIT):
    client_id = _get_client_id()
 
    try:
        resp = requests.get(
            'https://api-v2.soundcloud.com/search/tracks',
            params={'q': genre, 'client_id': client_id, 'limit': limit},
            headers={
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
                               '(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
                'Referer': 'https://soundcloud.com/',
                'Origin': 'https://soundcloud.com',
            },
            timeout=8,
        )
        resp.raise_for_status()
        data = resp.json()
        tracks = [_track_from_api(t) for t in data.get('collection', [])]
        return tracks if tracks else mock_songs_by_genre(genre)
    except Exception as e:
        print(f'[soundcloud_service] SoundCloud request failed ({_redact(e)}) -- falling back to mock songs')
        return mock_songs_by_genre(genre)